### PR TITLE
fix(workflow): Properly update project resources when transferring

### PR DESCRIPTION
**Pull request by barkbarkimashark**
* Conditionally create a new Environment by-name when transferring a project across orgs.
* Update SnubaQuery to reference the new Environment.

WOR-1663

### CODE REVIEW

File: src/sentry/models/project.py [hunk 1]
```diff
@@ -341,6 +342,11 @@ def transfer_to(self, team=None, organization=None):
             if not is_member:
                 rule.update(owner=None)
 
+        # null out environment_id so when the org(and its associated resources) eventually gets deleted
+        # the to_org resources aren't referencing to-be-deleted DB objects, blocking the deletion process
+        for alert_rule in list(alert_rules):
+            SnubaQuery.objects.filter(id=alert_rule.snuba_query_id).update(environment_id=None)
```
Comment (wedamija) [inline src/sentry/models/project.py:348]: When we look at how `Rule` objects are transferred over a few lines up it `get_or_create`s the environment in the new org and sets the environment id for the `Rule` to that environment. I think it makes sense to do the same here as well.

File: tests/sentry/models/test_project.py [hunk 2]
```diff
@@ -177,13 +178,36 @@ def test_transfer_to_organization_releases(self):
...
 
         ReleaseProjectEnvironment.objects.create(
             project=project, release=release, environment=environment
         )
 
+        assert Environment.objects.filter(id=environment.id).exists()
+        assert Environment.objects.filter(organization_id=from_org.id, projects=project).exists()
+        assert not Environment.objects.filter(
+            project_id=project.id
+        ).exists()  # project_id deprecated
```
Comment (wedamija) [inline tests/sentry/models/test_project.py:208]: This field is unused, let's skip testing it

File: tests/sentry/models/test_project.py [hunk 3]
```diff
@@ -212,32 +242,39 @@ def test_transfer_to_organization_alert_rules(self):
...
         )
+        snuba_query = SnubaQuery.objects.filter(id=alert_rule.snuba_query_id).get()
         rule1 = Rule.objects.create(label="another test rule", project=project, owner=team.actor)
         rule2 = Rule.objects.create(label="rule4", project=project, owner=from_user.actor)
 
         # should keep their owners
         rule3 = Rule.objects.create(label="rule2", project=project, owner=to_team.actor)
         rule4 = Rule.objects.create(label="rule3", project=project, owner=to_user.actor)
 
+        assert snuba_query
```
Comment (wedamija) [inline tests/sentry/models/test_project.py:262]: No need for this assert. If the `get` above has no rows, then it will raise.

File: tests/sentry/models/test_project.py [hunk 4]
```diff
@@ -212,32 +242,39 @@ def test_transfer_to_organization_alert_rules(self):
...
+        assert snuba_query.environment.id is environment.id
```
Comment (wedamija) [inline tests/sentry/models/test_project.py:263]: Generally use `==` unless you're testing that these are the exact same object.

File: src/sentry/models/project.py (hunk 1 above)
Comment (barkbarkimashark) [reply src/sentry/models/project.py]: Took a crack at your suggestion. Let me know what you think.

File: tests/sentry/tasks/test_deletion.py [hunk 5]
```diff
@@ -140,6 +146,83 @@ def test_handle_missing_record(self):
         assert not ScheduledDeletion.objects.filter(id=schedule.id).exists()
 
 
+class DeleteOrganizationTest(TestCase):
```
Comment (markstory) [inline tests/sentry/tasks/test_deletion.py:149]: There should be some existing organization tests in `tests/sentry/deletions/test_organization.py` should these go with the existing tests?

Comment (barkbarkimashark) [reply tests/sentry/tasks/test_deletion.py:149]: Ahh, I didn't see this before. Good suggestion! I think it does makes sense to move it here.

File: src/sentry/models/project.py [hunk 6]
```diff
@@ -341,6 +342,19 @@ def transfer_to(self, team=None, organization=None):
...
+        # conditionally create a new environment associated to the new Org -> Project -> AlertRule -> SnubaQuery
+        # this should take care of any potentially dead references from SnubaQuery -> Environment when deleting
+        # the old org
+        # alertrule ->  snuba_query -> environment_id
+        for snuba_id, environment_id in AlertRule.objects.fetch_for_project(self).values_list(
+            "snuba_query_id", "snuba_query__environment__id"
+        ):
+            SnubaQuery.objects.filter(id=snuba_id).update(
+                environment_id=Environment.get_or_create(
+                    self, name=environment_names.get(environment_id, None)
```
Comment (wedamija) [inline src/sentry/models/project.py:354]: `environment_names` will only contain environments that are used on `Rule` objects, since we build it based on `rules_by_environment_id`. If the `AlertRule` uses an environment that wasn't used on any `Rule` object then we'd fail to get the name here, probably worth covering that case too.

Comment (barkbarkimashark) [reply src/sentry/models/project.py]: Hmm, good catch. New commit should cover this scenario with the appropriate test.

File: src/sentry/deletions/__init__.py [hunk 7]
```diff
@@ -21,7 +21,7 @@
 --------------------
 
 The entrypoint into deletions for the majority of application code is via the ``ScheduledDeletion``
-model. This model lets you creation deletion jobs that are run in the future.
+model. This model lets you create deletion jobs that are run in the future.
```
Comment (wedamija) [inline src/sentry/deletions/__init__.py:24]: No need to change this time around, but generally try to keep unrelated changes into separate prs

File: src/sentry/models/project.py [hunk 8]
```diff
@@ -341,6 +342,31 @@ def transfer_to(self, team=None, organization=None):
...
 
+        # [Rule, AlertRule(SnubaQuery->Environment)]
+        # id -> name
+        environment_names_with_alerts = {
+            **environment_names,
+            **{
+                env_id: env_name
+                for env_id, env_name in AlertRule.objects.fetch_for_project(self).values_list(
+                    "snuba_query__environment__id", "snuba_query__environment__name"
+                )
```
Comment (wedamija) [inline src/sentry/models/project.py:353]: Nit: Could use `.distinct()` here to return fewer rows. Probably doesn't have a big impact.

**Review by wedamija — APPROVED** [review]

File: src/sentry/deletions/__init__.py (hunk 7 above)
Comment (barkbarkimashark) [reply src/sentry/deletions/__init__.py:24]: Will do for next time!
